docs/expression.md

```diff
@@ -0,0 +1,221 @@
+
+__EBNF Notation__
+```ebnf
+identifier_expression = name;
+```
+
+__Example__
+```ttr
+var a = 12
+a = 2
+
+// a is the identifier expression thats nested in a assignment expressions.
+
+val b = a / 7
+
+// a and b are both identifier expression that are nested within an expression tree.
+```
+
+## Literal expression
+
+Like identifiers, literals are expressions based on a single item. In this case a literal value like: 1, true/false or 'string'. These share a spot in the order of operations with identifiers under the combined name: terms. Like identifiers, these cannot exit on their own and are always nested within another expression.
+
+__EBNF Notation__
+```ebnf
+literal_expression = boolean | integer | string;
+```
+<sub>see: [primitives](primitives.md)</sub>
+
+__Example__
+```ttr
+var a = 12
+a = 2
+
+// 12 and 2 are both literal expressions.
+
+val b = a / 7
+
+// 7 is an literal expression.
+
+var c = true
+c = false
+
+// true and false are both literal expressions.
+
+var d = 'Niels'
+d = 'Marnix'
+
+// 'Niels' and 'Marnix' are both literal expressions.
+```
+
+## Binary expression 
+
+Used in performing an operation on two expressions. These are used to build the expression tree. The last leaf within the tree will always be a term (identifier or literal). Binary expression are used for mathematical operations, logic operations, assignment expressions and any valid combination between these. Binary expressions follow the [order of operations](#Order-of-opperations).
+
+Since binary expressions contain expressions for the left and right side of the operations it can go to a very deep nesting of expressions. In the example image below there is an example of how the binary expression tree is build.
+
+Expression: (a + b) * c + 7
+
+![Expression tree of the expression (a+b)*c+7](https://upload.wikimedia.org/wikipedia/commons/thumb/9/98/Exp-tree-ex-11.svg/500px-Exp-tree-ex-11.svg.png)
+
+<sub>Source: [wikimedia](https://commons.wikimedia.org/wiki/File:Exp-tree-ex-11.svg)</sub>
+
+For more information on how expression trees work and are build see the [wikipedia article](https://en.wikipedia.org/wiki/Binary_expression_tree) on binary expression tree's.
+
+__EBNF Notation__
+```ebnf
+binary_expression = expression, operator, expression;
+```
+
+<sub>see: [operators](operators.md)</sub>
+
+__Example__
+```ttr
+var a
+
+a = 2 * 3 
+
+// 2 * 3 
+// is a binary expression of 2 terms with an multiply operation.
+
+var b
+
+b = a / 1
+
+// a / 1 
+// is a binary expression of 2 terms with an division operation.
+
+val c = a * b + 12 / 2
+
+// a * b + 12 / 2 
+// is a binary expression of two binary expressions. When resolving these it will result as follows
+// binary expression left: a * b (Which is a also binary expressions)
+// operation: Add (+)
+// binary expression right 12 / 2 (Which is a also binary expressions)
+
+// The end result of c = 42
```

Comment:
```suggestion
val foo = 2 * 3
// "2 * 3" is a binary expression of 2 terms with a multiply operation.

val bar = foo / 1
// "foo / 1" is a binary expression of 2 terms with a division operation.

val baz = foo * bar + 12 / 2
// "foo * bar + 12 / 2" is a binary expression tree root. Both the left and right side contain another binary expression.
// The left side is "foo * bar" and the right is "12 / 2". Evaluating this will result in 42.
```